dali/ft/filecopy.cpp

```diff
@@ -2596,6 +2651,345 @@ void FileSprayer::pushParts()
     performTransfer();
 }
 
+using namespace Azure::Storage;
+using namespace Azure::Storage::Files;
+using namespace Azure::Storage::Blobs;
+
+enum ApiCopyStatus { Pending, Success, Failed, Aborted };
+interface IAPICopyClient : implements IInterface
+{
+    virtual ApiCopyStatus startCopy(const char *target, const char * source) = 0;
+    virtual ApiCopyStatus getProgress(CDateTime & dateTime, int64_t & outputLength) = 0;
+    virtual ApiCopyStatus abortCopy() = 0;
+    virtual ApiCopyStatus getStatus() const = 0;
+};
+
+class CApiCopyClient : implements IAPICopyClient, public CInterface
+{
+protected:
+    ApiCopyStatus status;
+public:
+    IMPLEMENT_IINTERFACE
+    virtual ApiCopyStatus getStatus() const
+    {
+        return status;
+    }
+    static void buildStorageURIBase(IStoragePlane * storagePlane, StringBuffer & uri)
+    {
+        const char *accountName = storagePlane->queryStorageApiAccount();
+        uri.appendf("https://%s", accountName);
+
+        if (storagePlane->getStorageType()==StorageType::StorageTypeAzureFile)
+        {
+            uri.append(".file");
+        }
+        else
+        {
+            dbgassertex(storagePlane->getStorageType()==StorageType::StorageTypeAzureBlob);
+            uri.append(".blob");
+        }
+        const char *container = storagePlane->queryStorageContainer();
+        uri.appendf(".core.windows.net/%s", container);
+    }
+    static void buildAzureStorageURI(IDistributedFile *distFile, unsigned partNum, const char * baseURI, const char *sasToken, StringBuffer & uri)
+    {
+        StringBuffer path;
+        distFile->queryPart(partNum).getStorageFilePath(path, 0);
+        StringBuffer tmp;
+        uri.appendf("%s%s%s", baseURI, encode_url(tmp, path.str()).str(), sasToken);
+    }
+};
+
+class AzureFileClient : public CApiCopyClient
+{
+    std::unique_ptr<Shares::ShareFileClient> fileClient;
+    std::unique_ptr<Shares::StartFileCopyOperation> fileCopyOp;
+public:
+    ApiCopyStatus startCopy(const char *target, const char * source)
+    {
+        try
+        {
+            fileClient.reset(new Shares::ShareFileClient(target));
+            fileCopyOp = std::make_unique<Shares::StartFileCopyOperation>(std::move(fileClient->StartCopy(source)));
+            status = ApiCopyStatus::Pending;
+        }
+        catch (const Azure::Core::RequestFailedException& e)
+        {
+            IERRLOG("AzureFileClient startCopy failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+            status = ApiCopyStatus::Failed;
+            throw MakeStringException(DFTERR_CopyFailed, "%s", e.ReasonPhrase.c_str());
+        }
+        return status;
+    }
+    ApiCopyStatus getProgress(CDateTime & dateTime, int64_t & outputLength)
+    {
+        if (status!=ApiCopyStatus::Pending)
```

Comment:
This does not initialise dateTime or outputLength - but the call on line 2950 passes uninitialised values, and assigns them unconditionally.  If nothing else would be better to clear them here...

Reply:
I've initialized dateTime and outputLength now.